### PR TITLE
Several improvements to `FiniteMap`.

### PR DESCRIPTION
**Pull request by ryzhyk**
See individual commits for details.

### CODE REVIEW

File: src/algebra/finite_map/mod.rs [hunk 1]
```diff
@@ -19,8 +22,14 @@ pub trait KeyProperties: 'static + Clone + Eq + Hash {}
 
 impl<T> KeyProperties for T where T: 'static + Clone + Eq + Hash {}
 
-// FIXME: this is not abstract enough.
-type SupportIter<'a, KeyType, ValueType> = hash_map::Keys<'a, KeyType, ValueType>;
+pub trait WithSupport<'a, KeyType>
```
Comment (mbudiu-vmw) [inline src/algebra/finite_map/mod.rs:25]: a comment would be nice.

File: src/algebra/finite_map/mod.rs [hunk 2]
```diff
@@ -39,16 +48,32 @@ where
     /// Find the value associated to the specified key
     fn lookup(&self, key: &Key) -> Value;
 
-    /// Return the set of values that are mapped to non-zero values.
-    fn support(&self) -> SupportIter<'_, Key, Value>;
+    /// Find the value associated to the specified key
+    ///
+    /// Return value `None` indicates that the key is not stored in the map,
+    /// and its associated value is `Value::zero()`.
+    fn get(&self, key: &Key) -> Option<&Value>;
```
Comment (mbudiu-vmw) [inline src/algebra/finite_map/mod.rs:55]: but this is wrong. FiniteMaps always map to a group, and if a value is not there it is effectively mapped to 0.

Comment (mbudiu-vmw) [reply src/algebra/finite_map/mod.rs:55]: FiniteMaps are total functions, so they always return a value.

File: src/algebra/finite_map/mod.rs [hunk 3]
```diff
@@ -178,8 +195,48 @@ where
         self.value.get(key).cloned().unwrap_or_else(Value::zero)
     }
 
-    fn support<'a>(&self) -> SupportIter<'_, Key, Value> {
-        self.value.keys()
+    fn get(&self, key: &Key) -> Option<&Value> {
```
Comment (mbudiu-vmw) [inline src/algebra/finite_map/mod.rs:198]: see above

File: src/algebra/finite_map/mod.rs (hunk 2 above)
Comment (mbudiu-vmw) [reply src/algebra/finite_map/mod.rs:55]: you can have a private helper function, but this should not be part of the trait.

File: src/algebra/finite_map/mod.rs [hunk 4]
```diff
@@ -178,8 +195,48 @@ where
...
+                let val = oe.get_mut();
+                f(val);
+                if val.is_zero() {
+                    oe.remove();
+                }
+            }
+            RawEntryMut::Vacant(ve) => {
+                let mut val = Value::zero();
+                f(&mut val);
+                ve.insert(key.clone(), val);
```
Comment (mbudiu-vmw) [inline src/algebra/finite_map/mod.rs:217]: we should not store val if it is zero

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: src/algebra/finite_map/mod.rs (hunk 2 above)
Comment (ryzhyk) [reply src/algebra/finite_map/mod.rs:55]: Renaming to `get_in_support`.

File: src/algebra/finite_map/mod.rs (hunk 4 above)
Comment (ryzhyk) [reply src/algebra/finite_map/mod.rs:217]: Right!  I am also fixing a similar issue in `ZSetHashMap::singleton()`.

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**